pyro/poutine/broadcast_messenger.py

```diff
@@ -0,0 +1,40 @@
+from __future__ import absolute_import, division, print_function
+
+from .messenger import Messenger
+
+
+class BroadcastMessenger(Messenger):
+    """
+    `BroadcastMessenger` automatically broadcasts the batch shape of
+    the stochastic function at a sample site when inside a single
+    or nested iarange context. The existing `batch_shape` must be
+    broadcastable with the size of the :class::`pyro.iarange`
+    contexts installed in the `cond_indep_stack`.
+    """
+    def _pyro_sample(self, msg):
+        """
+        :param msg: current message at a trace site.
+        """
+        if msg["done"] or msg["type"] != "sample":
+            return
+
+        dist = msg["fn"]
+        actual_batch_shape = getattr(dist, "batch_shape", None)
+        if actual_batch_shape is not None:
+            target_batch_shape = []
```

Comment:
Nice! I think this could be a little stricter and a little simpler if you used `-1` sizes, something like
```py
target_batch_shape = [-1 if size == 1 else size for size in actual_batch_shape]
for f in msg["conde_indep_stack"]:
    if f.dim is None:
        continue
    assert f.dim < 0
    if -f.dim > len(target_batch_shape):
        target_batch_shape = [-1] * (-f.dim - len(target_batch_shape)) + target_batch_shape
    elif target_batch_shape[f.dim] not in (-1, f.size):
       raise ValueError("... dim collision ...")
    target_batch_shape[f.dim] = f.size
    msg["fn"] = msg["fn"].expand(target_batch_shape)
```

Reply:
The problem that I was facing was with expanding smaller-sized latent sites when the iarange dims appear in a staggered fashion:

```python
with pyro.iarange("num_particles", 10, dim=-3):
    with pyro.iarange("components", 2, dim=-1):
         # with .expand([10, -1, 2]), we get s.shape == (10,)
         # with .expand([10, 1, 2]), we get s.shape == (10, 1, 2)
         s = pyro.sample("sample", dist.Bernoulli(0.5))    
```

Reply:
The snippet I suggested is order invariant. Note the final `.expand()` statement is outside of the loop.

Reply:
Just updated the snippet to clarify. The problem comes in the sample site in the second `iarange` where the outermost iarange is at -3 and the inner `components` one is at -1 (the `data` iarange is yet to come). If we expand `s` as `dist.Bernoulli(0.5).expand([10, -1, 2])` (note the default `-1`) it will give us `s.shape == torch.Size((10,))` whereas we want it to be of `torch.Size((10, 1, 2))` so that it can be correctly broadcasted by the `data` iarange.

Reply:
Thanks for explaining, I didn't know that -1 cannot be used for new dimensions in `torch.expand()`. Could we still try to catch expand errors early, while the `iarange.name` is still around? Maybe
```py
target_batch_shape = [None if size == 1 else size for size in actual_batch_shape]
for f in msg["cond_indep_stack"]:
    if f.dim is None:
        continue
    assert f.dim < 0
    if -f.dim > len(target_batch_shape):
        target_batch_shape = [None] * (-f.dim - len(target_batch_shape)) + target_batch_shape
    elif target_batch_shape[f.dim] not in (None, f.size):
        raise ValueError("Shape mismatch inside iarange('{}') at site {} dim {}, {} vs {}".format(
            f.name, msg['name'], f.dim, f.size, target_batch_shape[dim]))
    target_batch_shape[f.dim] = f.size
# ... remainder of your code ...
```

Reply:
Good point; will update.